### PR TITLE
`nn.Linear`: nD contiguous input + bias -- dispatch to addmm also when weight is sparse

### PR DESCRIPTION
**Pull request by nikitaved**
As per title.

It seems safe to be able to generalize to arbitrary contiguous inputs since `at::matmul` is likely to do the flattening to avoid `baddmm`.

Additionally, we guard for bias to be 1D and contiguous which is guaranteed to be fused with no copies.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -91,12 +108,9 @@ Tensor linear(const Tensor& input, const Tensor& weight, const std::optional<Ten
...
   if (bias->defined() && !input.is_xla()) {
-    // Also hit the fused path for contiguous 3D input, if not using xla
+    // Also hit the fused path for contiguous nD input, if not using xla
     // backend. Reshaping/flattening has some performance implications on xla.
-    bool is_contiguous = input.is_contiguous_or_false();
-    if (is_contiguous && input_dim == 3) {
-      return _flatten_nd_linear(input, weight, *bias);
-    } else if (is_contiguous && input.layout() == c10::kStrided && weight.layout() == c10::kStrided && bias->dim() == 1) {
+    if (input.is_contiguous_or_false()) {
       return _flatten_nd_linear(input, weight, *bias);
```
Comment (nikitaved) [inline aten/src/ATen/native/Linear.cpp:125]: The implicit assumption is that `weight` is 2D (1D `input` will be promoted to 2D). Do I need to guard for this?

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -50,18 +50,35 @@ static inline bool parseLinearFlatten3d() {
...
-    const auto input_sizes = input.sym_sizes();
-    // can't use -1 in reshape because it errors when a dimension is 0
-    c10::SymInt flattened_dim = 1;
-    for (int64_t i = 0, ndim = input_sizes.size(); i < ndim - 1; ++i) {
-      flattened_dim = flattened_dim * input_sizes[i];
+  const auto input_sizes = input.sym_sizes();
+
+  const auto result_flattened = [&]() -> Tensor {
+    const auto input_ncols = input_sizes.back();
+    const auto input_flattened_nrows = [&]() -> c10::SymInt {
```
Comment (eellison) [inline aten/src/ATen/native/Linear.cpp:57]: nit: do we need nested, lambdas for `input_flattened_nrows` which is only invoked once ? this feels unnecessary. or maybe i am just not reading as much c++ these days.

File: aten/src/ATen/native/Linear.cpp [hunk 3]
```diff
@@ -90,15 +107,23 @@ Tensor linear(const Tensor& input, const Tensor& weight, const std::optional<Ten
...
+  const auto is_bias_likely_fusable = (
+      bias->defined() &&
+      // cuBLASLt: will fuse in the epilogue without copies
+      // when input/weight/bias are all strided.
+      // When weight is not strided, bias will not be fused,
+      // but we can still dispatch here to avoid at::matmul
+      // path which will probably use a very similar
+      // flattening optimization.
+      (bias->dim() == 1 && bias->is_contiguous_or_false())
+  );
```
Comment (eellison) [inline aten/src/ATen/native/Linear.cpp:120]: This is the generic, linear lowering. it feels a bit strange to me to add cublasLT logic here. Should we at least check this is a cuda tensor? or reason not to?

**Review by eellison — COMMENTED** [review]
Seems like a good change, cc @ngimel for thoughts as well. nit: add test ?

File: aten/src/ATen/native/Linear.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/Linear.cpp:57]: This seems like a style preference, maybe? We deployed these quite a lot in torch.linalg to enforce const lvalues as much as possible and to reduce namespace pollution. Are there objectives against it? I can change.

File: aten/src/ATen/native/Linear.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/Linear.cpp:120]: This logic was here before (although in the later branch). I decided to make it a bit more explicit as to why. Only dim=1 bias guarantees a fused kernel in CUDA, though ...

File: aten/src/ATen/native/Linear.cpp (hunk 2 above)
Comment (eellison) [reply aten/src/ATen/native/Linear.cpp:57]: I'll defer to you.

**Review by ngimel — APPROVED** [review]
